File: sdks/go/mcp/layer2.go
package mcp

import (
	"context"
	"encoding/json"
	"fmt"

	"github.com/solvapay/solvapay-go/internal/nativecall"
)

func callLayer2(ctx context.Context, fn string, args any) (json.RawMessage, error) {
	raw, err := json.Marshal(args)
	if err != nil {
		return nil, err
	}
	value, err := nativecall.CallValueJSON(ctx, fn, string(raw))
	if err != nil {
		return nil, err
	}
	if len(value) == 0 || string(value) == "null" {
		return nil, fmt.Errorf("native %s returned empty value", fn)
	}
	return value, nil
}

func paywallToolResult(ctx context.Context, message string, gate json.RawMessage) (json.RawMessage, error) {
	if len(gate) == 0 {
		gate = json.RawMessage(`{}`)
	}
	type args struct {
		Message           string          `json:"message"`
		StructuredContent json.RawMessage `json:"structuredContent"`
	}
	return callLayer2(ctx, "sv_paywall_tool_result_binding", args{Message: message, StructuredContent: gate})
}

func makeResponseResult(ctx context.Context, data json.RawMessage, options map[string]any, emitted []json.RawMessage) (json.RawMessage, error) {
	payload := map[string]any{"data": data}
	if options != nil {
		payload["options"] = options
	}
	if len(emitted) > 0 {
		payload["emittedBlocks"] = emitted
	}
	return callLayer2(ctx, "sv_make_response_result_binding", payload)
}

func assertResponseResult(ctx context.Context, value json.RawMessage) (json.RawMessage, error) {
	return callLayer2(ctx, "sv_assert_response_result_binding", map[string]any{"value": value})
}

func buildPayableToolResult(ctx context.Context, envelope json.RawMessage) (json.RawMessage, error) {
	return callLayer2(ctx, "sv_build_payable_tool_result_binding", map[string]any{"envelope": envelope})
}

File: sdks/ruby-mcp/lib/solvapay/mcp/layer2.rb
# frozen_string_literal: true

require "solvapay"

module SolvaPay
  module Mcp
    module Layer2
      class << self
        def paywall_tool_result(message, gate)
          as_object_map(
            SolvaPay::NativeDispatch.call_sync(
              "paywall_tool_result",
              { "message" => message, "structuredContent" => gate },
            ),
          )
        end

        def make_response_result(data, options, emitted_blocks)
          args = { "data" => data }
          args["options"] = options unless options.nil?
          args["emittedBlocks"] = emitted_blocks unless emitted_blocks.empty?
          as_object_map(SolvaPay::NativeDispatch.call_sync("make_response_result", args))
        end

        def assert_response_result(value)
          as_object_map(SolvaPay::NativeDispatch.call_sync("assert_response_result", { "value" => value }))
        end

        def build_payable_tool_result(envelope)
          as_object_map(
            SolvaPay::NativeDispatch.call_sync("build_payable_tool_result", { "envelope" => envelope }),
          )
        end

        def as_object_map(value)
          return value if value.is_a?(Hash)

          raise SolvaPay::SolvaPayError, "native call returned unexpected value"
        end
      end
    end
  end
end
